internal/pkg/warewulfd/util.go

```diff
@@ -1,51 +1,82 @@
 package warewulfd
 
 import (
-	"io"
 	"net/http"
 	"os"
-	"strconv"
 
+	nodepkg "github.com/hpcng/warewulf/internal/pkg/node"
+	"github.com/hpcng/warewulf/internal/pkg/overlay"
+	"github.com/hpcng/warewulf/internal/pkg/util"
 	"github.com/hpcng/warewulf/internal/pkg/wwlog"
-	"github.com/pkg/errors"
 )
 
-func sendFile(w http.ResponseWriter, filename string, sendto string) error {
+func sendFile(
+	w http.ResponseWriter,
+	req *http.Request,
+	filename string,
+	sendto string) error {
+
 	fd, err := os.Open(filename)
 	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
 		return err
 	}
 	defer fd.Close()
 
-	FileHeader := make([]byte, 512)
-	_, err = fd.Read(FileHeader)
+	stat, err := fd.Stat()
 	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
-		return errors.Wrap(err, "failed to read header")
+		return err
 	}
 
-	FileContentType := http.DetectContentType(FileHeader)
-	FileStat, _ := fd.Stat()
-	FileSize := strconv.FormatInt(FileStat.Size(), 10)
+	http.ServeContent(
+		w,
+		req,
+		filename,
+		stat.ModTime(),
+		fd )
 
-	_, err = fd.Seek(0, 0)
-	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		return errors.Wrap(err, "failed to seek")
-	}
+	wwlog.Send("%15s: %s", sendto, filename)
 
-	w.Header().Set("Content-Disposition", "attachment; filename=kernel")
-	w.Header().Set("Content-Type", FileContentType)
-	w.Header().Set("Content-Length", FileSize)
+	return nil
+}
 
-	_, err = io.Copy(w, fd)
-	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		return errors.Wrap(err, "failed to copy")
+func getOverlayFile(
+	nodeId string,
+	stage_overlays []string,
+	autobuild bool ) (stage_file string, err error) {
+
+	stage_file = overlay.OverlayImage(nodeId, stage_overlays)
+	err = nil
+
+	build := !util.IsFile(stage_file)
+
+	if !build && autobuild {
+		build = util.PathIsNewer(stage_file, nodepkg.ConfigFile)
+
+		for _, overlayname := range stage_overlays {
+			build = build || util.PathIsNewer(stage_file, overlay.OverlaySourceDir(overlayname))
+		}
 	}
 
-	wwlog.Send("%15s: %s", sendto, filename)
+	if build {
+		wwlog.Serv("BUILD %15s, overlays %v", nodeId, stage_overlays)
+
+		args := []string{"overlay", "build"}
+
+		for _, overlayname := range stage_overlays {
+			args = append(args, "-O", overlayname)
+		}
+
+		args = append(args, nodeId)
+
+		out, err := util.RunWWCTL(args...)
```

Comment:
I think in the future we should try to call  `overlay.BuildOverlayIndir(nodeId, stage_overlays, OverlayDir)` directly. The _OverlayDir_ is technical dept which we/I have to fix. With #429 there is a start of refactoring of the overlay handling and get rid of _ChDir_ which triggered this.
But given the late stage of development lets go with this.